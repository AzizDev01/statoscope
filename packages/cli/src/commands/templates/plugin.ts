import { concatTemplateParts, templatePartsByOptions } from './templateParts';
import { FileExt, RuleTemplatesPartsFromFileExt, Template } from './types';

export const pluginTemplate: Template = (templateOptions) => {
  const plugin = templatePartsByOptions(templateOptions, {
    js: jsPlugin,
    ts: tsPlugin,
  });

  return concatTemplateParts(plugin);
};

const jsPlugin: RuleTemplatesPartsFromFileExt[FileExt.js] = {
  imports: {
    commonjs: `const asyncChunkRule = require('./rule.statoscope.js')`,
    esm: `import asyncChunkRule from './rule.statoscope.js';`,
  },
  body: `/**
  * @typedef {import('@statoscope/webpack-model').Prepared} Prepared
  * @returns {import('@statoscope/stats-validator/dist/plugin').PluginDescription<Prepared>} PluginDescription
 */
const newStatoscopePlugin = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: {
    commonjs: 'module.exports = newStatoscopePlugin',
    esm: 'export default newStatoscopePlugin',
  },
};

const tsPlugin: RuleTemplatesPartsFromFileExt[FileExt.ts] = {
  imports: `import type { PluginFn } from '@statoscope/stats-validator/dist/plugin';
import type { Prepared } from '@statoscope/webpack-model';
import asyncChunkRule from './rule.statoscope.ts';`,
  body: `
const newStatoscopePlugin: PluginFn<Prepared> = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: 'export default newStatoscopePlugin',
};
