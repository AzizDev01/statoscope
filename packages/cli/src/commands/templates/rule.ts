export function ruleTemplate(): string {
  return `import { Rule } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';

//Statoscope Plugins examples:
https://github.com/statoscope/statoscope/tree/c9d0c148c1e38a87696d79485f37e598110db80c/packages/stats-validator-plugin-webpack#rules


${asyncChunkRule}

export default asyncChunkRule;`;
}

export const asyncChunkRule = `
type WebpackRule<TParams> = Rule<TParams, Prepared>;

type Params = string[];

const asyncChunkRule: WebpackRule<Params> = (ruleParams, data, api): void => {
  const normalizedParams: string[] = ruleParams ? ruleParams : [];
  const chunks = Array.from(new Set(...data.compilations.map((c) => c.chunks)));
  chunks
    .filter((c) =>
      c.names.some((n) => normalizedParams.includes(n) && c.initial !== false)
    )
    .forEach((f) => {
      api.message(f.names[0] + ' is not async chunk.');
    });
};`;
