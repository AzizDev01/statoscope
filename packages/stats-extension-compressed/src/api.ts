import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Resource, Size } from './generator';

export type API = (compilationId: string, resourceId: string) => Size | null;

const makeAPI: APIFactory<Format, API> = (source) => {
  const sizeResolvers: Map<string, Resolver<string, Resource>> = new Map();

  for (const compilation of source.payload.compilations) {
    sizeResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.resources, (r) => r.id)
    );
  }

  return (compilationId: string, resourceId: string): Size | null => {
    return sizeResolvers.get(compilationId)?.(resourceId)?.size ?? null;
  };
};

export default makeAPI;
