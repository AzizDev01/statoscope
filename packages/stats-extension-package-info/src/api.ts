import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Package, Instance } from './generator';

export type API = {
  getPackage: (compilationId: string, packageName: string) => Package | null;
  getInstance: (
    compilationId: string,
    packageName: string,
    instancePath: string
  ) => Instance | null;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const packageResolvers: Map<string, Resolver<string, Package>> = new Map();
  const instanceResolvers: Map<Package, Resolver<string, Instance>> = new Map();

  for (const compilation of source.payload.compilations) {
    packageResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.packages, (r) => r.name)
    );

    for (const packageItem of compilation.packages) {
      instanceResolvers.set(
        packageItem,
        makeEntityResolver(packageItem.instances, (r) => r.path)
      );
    }
  }

  return {
    getPackage(compilationId: string, packageId: string): Package | null {
      return packageResolvers.get(compilationId)?.(packageId) ?? null;
    },
    getInstance(
      compilationId: string,
      packageId: string,
      instancePath: string
    ): Instance | null {
      const resolvedPackage = packageResolvers.get(compilationId)?.(packageId) ?? null;

      if (!resolvedPackage) {
        return null;
      }

      return instanceResolvers.get(resolvedPackage)?.(instancePath) ?? null;
    },
  };
};

export default makeAPI;
