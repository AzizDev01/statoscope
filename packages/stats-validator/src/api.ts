import {
  API,
  MakeAPIParams,
  RuleDescriptor,
  Storage,
} from '@statoscope/types/types/validation';

export function makeAPI(params?: MakeAPIParams): API {
  const storage: Storage = [];
  let hasErrors = false;
  let errors = 0;
  let warnings = 0;
  let infos = 0;
  let descriptor: RuleDescriptor | null = null;
  return {
    setRuleDescriptor(desc: RuleDescriptor): void {
      descriptor = desc;
    },
    getRuleDescriptor(): RuleDescriptor | null {
      return descriptor;
    },
    hasErrors(): boolean {
      return hasErrors;
    },
    getInfoTotal(): number {
      return infos;
    },
    getWarnTotal(): number {
      return warnings;
    },
    getErrorTotal(): number {
      return errors;
    },
    getStorage(): Storage {
      return storage;
    },
    warn(message, filenameOrOptions): void {
      if (params?.warnAsError) {
        hasErrors = true;
      }

      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      storage.push({
        type: 'warn',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        related: filenameOrOptions?.related,
        message,
      });
      warnings++;
    },
    error(message, filenameOrOptions): void {
      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      hasErrors = true;
      storage.push({
        type: 'error',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        related: filenameOrOptions?.related,
        message,
      });
      errors++;
    },
    info(message, filenameOrOptions): void {
      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      storage.push({
        type: 'info',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        related: filenameOrOptions?.related,
        message,
      });
      infos++;
    },
  };
}
