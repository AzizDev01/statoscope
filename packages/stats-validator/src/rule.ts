import { API } from './api';

export type RuleDataInput<TInput> = { reference?: TInput; input: TInput };

export type Rule<TParams, TInput> = (
  params: TParams | null,
  data: RuleDataInput<TInput>,
  api: API
) => void;
