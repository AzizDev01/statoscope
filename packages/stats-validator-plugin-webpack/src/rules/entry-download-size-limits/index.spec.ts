import { API, makeAPI } from '@statoscope/stats-validator/dist/api';
import { RuleDataInput } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';
import plugin from '../..';
import statsV5 from '../../../../../test/bundles/simple/stats-prod.json';
import rule, { Params } from './';

function prepareAPI(ruleParams: Params): API {
  const pluginInstance = plugin();
  const prepared: RuleDataInput<Prepared> = pluginInstance.prepare!([
    { name: 'input.json', data: statsV5 },
  ]);
  const api = makeAPI({ warnAsError: false });

  rule(ruleParams, prepared, api);

  return api;
}

type Case = {
  name: string;
  params: Params;
};

const maxSizeCases: Case[] = [
  {
    name: 'global',
    params: {
      global: {
        maxSize: 100,
      },
    },
  },
  {
    name: 'override global',
    params: {
      global: {
        maxSize: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxSize: 100,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      global: {
        maxSize: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      global: {
        maxSize: 100,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxSize: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      useCompressedSize: false,
      global: {
        maxSize: 100,
      },
    },
  },
];

const maxInitialSizeCases: Case[] = [
  {
    name: 'global',
    params: {
      global: {
        maxInitialSize: 100,
      },
    },
  },
  {
    name: 'override global',
    params: {
      global: {
        maxInitialSize: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxInitialSize: 100,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      global: {
        maxInitialSize: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      global: {
        maxInitialSize: 100,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxInitialSize: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      useCompressedSize: false,
      global: {
        maxInitialSize: 100,
      },
    },
  },
];

const maxAsyncSizeCases: Case[] = [
  {
    name: 'global',
    params: {
      global: {
        maxAsyncSize: 100,
      },
    },
  },
  {
    name: 'override global',
    params: {
      global: {
        maxAsyncSize: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxAsyncSize: 100,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      global: {
        maxAsyncSize: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      global: {
        maxAsyncSize: 100,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxAsyncSize: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      useCompressedSize: false,
      global: {
        maxAsyncSize: 100,
      },
    },
  },
];

const excludeCases: Case[] = [
  {
    name: 'entry string',
    params: {
      exclude: ['two'],
      global: {
        maxSize: 1,
      },
    },
  },
  {
    name: 'entry regexp',
    params: {
      exclude: [/tw/],
      global: {
        maxSize: 1,
      },
    },
  },
  {
    name: 'entry object',
    params: {
      exclude: [{ type: 'entry', name: /tw/ }],
      global: {
        maxSize: 1,
      },
    },
  },
  {
    name: 'no exclude',
    params: {
      exclude: [/foooo/],
      global: {
        maxSize: 1,
      },
    },
  },
];

describe('maxSize', () => {
  test.each(maxSizeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('maxInitialSize', () => {
  test.each(maxInitialSizeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('maxAsyncSize', () => {
  test.each(maxAsyncSizeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('exclude', () => {
  test.each(excludeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});
