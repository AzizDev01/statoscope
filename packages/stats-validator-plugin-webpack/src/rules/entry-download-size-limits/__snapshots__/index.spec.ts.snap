// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exclude entry object 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 1,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.44 kb. It's over the 0.00 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`exclude entry regexp 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 1,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.44 kb. It's over the 0.00 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`exclude entry string 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 1,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.44 kb. It's over the 0.00 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`exclude no exclude 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 1,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.44 kb. It's over the 0.00 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 1,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of assets is 8.53 kb. It's over the 0.00 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncSize global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": 100,
              "maxInitialSize": null,
              "maxSize": null,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of async assets is 0.13 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncSize not match - global 1`] = `Array []`;

exports[`maxAsyncSize not match - override global 1`] = `Array []`;

exports[`maxAsyncSize override global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": 100,
              "maxInitialSize": null,
              "maxSize": null,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of async assets is 0.13 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncSize with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 142,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 12253,
            "rule": Object {
              "maxAsyncSize": 100,
              "maxInitialSize": null,
              "maxSize": null,
            },
            "size": 12395,
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of async assets is 0.14 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialSize global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": 100,
              "maxSize": null,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of initial assets is 0.44 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": 100,
              "maxSize": null,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of initial assets is 8.40 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialSize not match - global 1`] = `Array []`;

exports[`maxInitialSize not match - override global 1`] = `Array []`;

exports[`maxInitialSize override global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": 100,
              "maxSize": null,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of initial assets is 0.44 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": 100,
              "maxSize": null,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of initial assets is 8.40 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialSize with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 799,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": 100,
              "maxSize": null,
            },
            "size": 799,
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of initial assets is 0.78 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 142,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 12253,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": 100,
              "maxSize": null,
            },
            "size": 12395,
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of initial assets is 11.97 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxSize global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 100,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.44 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 100,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of assets is 8.53 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxSize not match - global 1`] = `Array []`;

exports[`maxSize not match - override global 1`] = `Array []`;

exports[`maxSize override global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 449,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 100,
            },
            "size": 449,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.44 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 135,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 8604,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 100,
            },
            "size": 8739,
            "useCompressedSize": undefined,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of assets is 8.53 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxSize with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 0,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "one",
            "initialSize": 799,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 100,
            },
            "size": 799,
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size of assets is 0.78 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "asyncSize": 142,
            "compilation": "718159f8e48dcecd92c9",
            "entry": "two",
            "initialSize": 12253,
            "rule": Object {
              "maxAsyncSize": null,
              "maxInitialSize": null,
              "maxSize": 100,
            },
            "size": 12395,
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              size: #.size,
              initialSize: #.initialSize,
              asyncSize: #.asyncSize,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size of assets is 12.10 kb. It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;
